src/ast.rs

```diff
@@ -201,111 +210,171 @@ impl fmt::Display for Diagnostic {
   }
 }
 
-pub(crate) struct ParsedModule {
+/// A parsed module with comments.
+pub trait ParsedAst {
+  /// Gets the comments found in the source file.
+  fn comments(&self) -> &dyn swc_common::comments::Comments;
+  /// Gets the parsed module—AST.
+  fn module(&self) -> &swc_ecmascript::ast::Module;
+  /// Gets a position from the specified byte position.
+  /// Note: This should/will panic when the byte position is
+  // outside the bounds of the source file.
+  fn get_position(&self, pos: BytePos) -> Position;
+}
+
+/// Parses modules.
+pub trait AstParser {
+  /// Parses the provided information to a `ParsedAst` and returns a reference.
+  fn parse(
+    &mut self,
+    specifier: &ModuleSpecifier,
+    source: &str,
+    media_type: MediaType,
+  ) -> Result<&dyn ParsedAst, Diagnostic>;
+}
+
+pub struct DefaultParsedAst {
   comments: SingleThreadedComments,
   module: Module,
   text_lines: TextLines,
 }
 
-impl ParsedModule {
-  pub fn analyze_dependencies(&self) -> Vec<DependencyDescriptor> {
-    analyze_dependencies(&self.module, &self.comments)
-      .into_iter()
-      .filter(|desc| desc.kind != DependencyKind::Require)
-      .collect()
+impl ParsedAst for DefaultParsedAst {
+  fn comments(&self) -> &dyn swc_common::comments::Comments {
+    &self.comments
   }
 
-  pub fn analyze_deno_types(
-    &self,
-    desc: &DependencyDescriptor,
-  ) -> Option<(String, Range)> {
-    let comment = desc.leading_comments.last()?;
-    let captures = DENO_TYPES_RE.captures(&comment.text)?;
-    if let Some(m) = captures.get(1) {
-      Some((
-        m.as_str().to_string(),
-        Range::from_comment_match(comment, self, &m),
-      ))
-    } else if let Some(m) = captures.get(2) {
-      Some((
-        m.as_str().to_string(),
-        Range::from_comment_match(comment, self, &m),
-      ))
-    } else {
-      unreachable!("Unexpected captures from deno types regex")
-    }
+  fn module(&self) -> &swc_ecmascript::ast::Module {
+    &self.module
+  }
+
+  fn get_position(&self, pos: BytePos) -> Position {
+    Position::from_pos(&self.text_lines, pos)
   }
+}
+
+/// Get the module's leading comments, where triple slash directives might
+/// be located.
+pub(crate) fn get_leading_comments(parsed_ast: &dyn ParsedAst) -> Vec<Comment> {
+  parsed_ast
+    .comments()
+    .get_leading(parsed_ast.module().span.lo)
+    .unwrap_or_else(Vec::new)
+}
```

Comment:
I believe we use this functionality heavily in `deno_doc` and `deno_lint`, why not make it `pub`?

Reply:
I'll move it to the trait.

Reply:
Addressed in ac3853372c1ef6b76d7414c815c93a8681818aca